_posts/2020-04-09-flink-serialization-tuning-vol-1.md

```diff
@@ -0,0 +1,277 @@
+
+Since serialization is so crucial to your Flink job, we would like to highlight Flink’s serialization stack in a series of blog posts starting with looking at the different ways Flink can serialize your data types.
+
+{% toc %}
+
+# Recap: Flink Serialization
+
+Flink handles [data types and serialization](https://ci.apache.org/projects/flink/flink-docs-release-1.10/dev/types_serialization.html) with its own type descriptors, generic type extraction, and type serialization framework. We recommend reading through the [documentation](https://ci.apache.org/projects/flink/flink-docs-release-1.10/dev/types_serialization.html) first in order to be able to follow the arguments we present below. In essence, Flink tries to infer information about your job’s data types for wire and state serialization, and to be able to use grouping, joining, and aggregation operations by referring to individual field names, e.g. 
+`stream.keyBy(“ruleId”)` or 
+`dataSet.join(another).where("name").equalTo("personName")`. It also allows optimizations in the serialization format as well as reducing unnecessary de/serializations (mainly in certain Batch operations as well as in the SQL/Table APIs).
+
+
+# Choice of Serializer
+
+Apache Flink's out-of-the-box serialization can be roughly divided into the following groups:
+
+- **Flink-provided special serializers** for basic types (Java primitives and their boxed form), arrays, composite types (tuples, Scala case classes, Rows), and a few auxiliary types (Option, Either, Lists, Maps, …),
+
+- **POJOs**; a public, standalone class with a public no-argument constructor and all non-static, non-transient fields in the class hierarchy either public or with a public getter- and a setter-method; see [POJO Rules](https://ci.apache.org/projects/flink/flink-docs-release-1.10/dev/types_serialization.html#rules-for-pojo-types),
+
+- **Generic types**; user-defined data types that are not recognized as a POJO and then serialized via [Kryo](https://github.com/EsotericSoftware/kryo).
+
+Alternatively, you can also register [custom serializers](https://ci.apache.org/projects/flink/flink-docs-release-1.10/dev/custom_serializers.html) for user-defined data types. This includes writing your own serializers or integrating other serialization systems like [Google Protobuf](https://developers.google.com/protocol-buffers/) or [Apache Thrift](https://thrift.apache.org/) via [Kryo](https://github.com/EsotericSoftware/kryo). Overall, this gives quite a number of different options of serializing user-defined data types and we will elaborate 7 of them in the sections below.
```

Comment:
Have you considered replacing 7 by "seven"?
https://www.grammarly.com/blog/when-to-spell-out-numbers/

Reply:
I usually apply that rule for lower numbers, let's say 1,2,3, but sure, why not